### RANDOM FOREST REGRESSION

Use Random Forest when you have a lot of variables or features(columns).

It is bassically used for both classification and regression problems.

Using a random forest is almost pointless with only two dependent variables because a strength in the random forest is using a random subset of variables for each node split when determining which to use.

It is commonly used in stock market; predictt profit and loss by identifying stock behaviour.
consumer finance survey
calculation of harmful gases



In [171]:
# Import the libraries needed

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestRegressor

In [172]:
# import the dataset
dataset = pd.read_csv('/Users/lambertagunbiade/Desktop/SUPERVISED MACHINE LEARNING /HousePrice2.csv')
dataset.head(4)

Area  No of Rooms  Age of House  Price 
0  2422            4            35   32500
1  2130            3            39   27000
2  1900            2            16   24000
3  3100            6            41   44000

* First 4 rows of our data.

#### Separate the target variable from the predictors

In [173]:
X = dataset.iloc[0:,0:-1].values
y = dataset.iloc[0:,-1].values

#### Split data into training and testing data

In [174]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

#### Instantiate the random forest regression class

In [175]:
rand_regressor = RandomForestRegressor(n_estimators=15)
rand_regressor.fit(X_train,y_train)

RandomForestRegressor(n_estimators=15)

The decision for the value of n_estimator of our model was because we got the best estimate at this value of n-estimators. However, we know that the higher the number of estimators as well as more number of iteration of our model, the better its performance and accuracy.


credit: https://www.sciencedirect.com/science/article/pii/S1877050922001016

### Model Prediction

In [176]:
# training data 
y_pred_rand_train = rand_regressor.predict(X_train)

#testing data
y_pred_rand_test = rand_regressor.predict(X_test)

In [177]:
print('Actual prices of the land\n\n', y_test,'\n\n Predicted prices by Random Forest regression model\n',y_pred_rand_test.round(2))


# print('\nPredicted prices by the linear regression model\n\n', y_pred_rand.round(2))

Actual prices of the land

 [27000 51000 43000 47500] 

 Predicted prices by Random Forest regression model
 [29592.67 48633.33 43066.67 48360.  ]


The predicted charges above seem to have a huge difference when compared with the actual charges.
Let us see the difference between the real and predicted charges

In [178]:
prediction_difference = y_pred_rand_test - y_test
prediction_difference.round(2)

array([ 2592.67, -2366.67,    66.67,   860.  ])

Although the model variation is from -2366.67 to 2592.67.0 but its performance looks great. Let us see its R Squared values.

#### R SQUARED VALUES

In [179]:
# Import r sqaured library and create an object of it thus;
from sklearn.metrics import r2_score

In [180]:
# Training data r square value
r_square_rand_train = r2_score(y_train, y_pred_rand_train)

# Testing data r square value
r_square_rand_test = r2_score(y_test, y_pred_rand_test)

print('R square values:- goodness of fit')
print('Training data: ', r_square_rand_train.round(2),'Testing data: ',r_square_rand_test.round(2) )

R square values:- goodness of fit
Training data:  0.96 Testing data:  0.96


The random forest regression model has performed greatly with a knowlege of variation of the training and test data to be 96% and 96% respectively. Thus, it is reliable for us to predict house prices  of unseen data with the three features in our dataset.

#### LASSO REGRESSION

In [181]:
# import the ridge regression class
from sklearn.linear_model import Lasso

##### Instantiate and fit the model on the training data

In [182]:
lass = Lasso(alpha = 2) # Note: The choice of alpha is quite sensitive.
lass.fit(X_train, y_train)

Lasso(alpha=2)

#### MODEL PREDICTION

In [183]:
# Training data prediction
y_pred_lasso_train = lass.predict(X_train)

# Testing data prediction
y_pred_lasso_test = lass.predict(X_test)

In [184]:
print('Actual prices of the land\n\n', y_test,'\n\n Predicted prices by Random Forest regression model\n',y_pred_lasso_test.round(2))


# print('\nPredicted prices by the linear regression model\n\n', y_pred_rand.round(2))

Actual prices of the land

 [27000 51000 43000 47500] 

 Predicted prices by Random Forest regression model
 [28284.06 52330.23 39387.41 46129.95]


#### Prediction difference

In [188]:
prediction_difference = y_pred_lasso_test - y_test
prediction_difference.round(2)

array([ 1284.06,  1330.23, -3612.59, -1370.05])

This model gave a prediction difference of the predicted prices from the actual land prices to be -3612.59 to 1330.23
This range is definitely more further apart compared with that of random forest.

#### R SQUARED VALUES

In [190]:
# Training data r square value
lasso_r_square_train = r2_score(y_train, y_pred_lasso_train)

# Testing data r square value
lasso_r_square_test = r2_score(y_test, y_pred_lasso_test)
      
print('R square values:- goodness of fit')
print('Training data: ', lasso_r_square_train.round(2),'Testing data: ',lasso_r_square_test.round(2) )

R square values:- goodness of fit
Training data:  0.97 Testing data:  0.95


The R Squared value has shown a great knowledge on the variation of the unseen data. However, the random forest regression model should be adopted for this dataset because, it has a close difference in the price predicted. Also in the variation between the test and training data. 